File: nodejs/test/worker-causality.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { readFileSync } from "node:fs";
import { PassThrough } from "node:stream";
import {
    createMessageConnection,
    StreamMessageReader,
    StreamMessageWriter,
} from "vscode-jsonrpc/node.js";
import { describe, expect, it, onTestFinished } from "vitest";
import { CopilotSession } from "../src/session.js";
import type { SessionEvent } from "../src/index.js";

type Wire = Record<string, unknown>;
const corpus = JSON.parse(
    readFileSync(new URL("../../test/worker-causality.json", import.meta.url), "utf8")
) as {
    valid: { name: string; event?: Wire; result?: Wire }[];
    invalid: { name: string; value: unknown }[];
    boundaries: { name: string; value: unknown; accepted: boolean }[];
    workflowCompleted: { event: Wire };
};

function payload(value: Wire, event: boolean): Wire {
    return event ? (value.data as Wire) : value;
}

function readers() {
    const requests = new PassThrough();
    const responses = new PassThrough();
    const client = createMessageConnection(
        new StreamMessageReader(responses),
        new StreamMessageWriter(requests)
    );
    const server = createMessageConnection(
        new StreamMessageReader(requests),
        new StreamMessageWriter(responses)
    );
    let reply: unknown;
    server.onRequest(() => reply);
    client.listen();
    server.listen();
    const session = new CopilotSession("codec-session", client);
    const observed: SessionEvent[] = [];
    const unsubscribe = session.on((event) => observed.push(event));
    onTestFinished(() => {
        unsubscribe();
        client.dispose();
        server.dispose();
        requests.destroy();
        responses.destroy();
    });
    return {
        async read(wire: Wire, event: boolean): Promise<Wire> {
            if (event) {
                session._dispatchEvent(wire as unknown as SessionEvent);
                return observed.at(-1) as unknown as Wire;
            }
            reply = wire;
            return (await session.rpc.tasks.sendMessage({
                id: "worker",
                message: "test",
            })) as unknown as Wire;
        },
        async history(wire: Wire): Promise<SessionEvent[][]> {
            reply = { events: [wire], cursor: "cursor", hasMore: false, cursorStatus: "ok" };
            return [await session.getEvents(), (await session.rpc.eventLog.read({})).events];
        },
    };
}

describe("worker causality public readers", () => {
    it.each(corpus.valid)(
        "$name preserves exact observations and historical product bytes",
        async (test) => {
            const reader = readers();
            const event = test.event !== undefined;
            const wire = structuredClone((test.event ?? test.result) as Wire);
            const original = JSON.stringify(wire);
            expect(payload(await reader.read(wire, event), event).workerCausality).toEqual(
                payload(wire, event).workerCausality
            );
            expect(JSON.stringify(wire)).toBe(original);

            delete payload(wire, event).workerCausality;
            const baseline = JSON.stringify(await reader.read(wire, event));
            for (const invalid of corpus.invalid) {
                const candidate = structuredClone(wire);
                payload(candidate, event).workerCausality = invalid.value;
                expect(JSON.stringify(await reader.read(candidate, event)), invalid.name).toBe(
                    baseline
                );
                if (event) {
                    for (const events of await reader.history(candidate)) {
                        expect(JSON.stringify(events[0]), invalid.name).toBe(baseline);
                    }
                }
            }
        }
    );

    it.each(corpus.boundaries)(
        "$name enforces compact UTF-8 bytes and tolerates unknown fields",
        async (test) => {
            const reader = readers();
            const wire = structuredClone(corpus.valid[0].event as Wire);
            payload(wire, true).workerCausality = test.value;
            const data = payload(await reader.read(wire, true), true);
            expect("workerCausality" in data).toBe(test.accepted);
            expect(data.content).toBe(payload(wire, true).content);
        }
    );

    it("decodes the canonical workflow completion notification with typed fields", async () => {
        const event = corpus.workflowCompleted.event as unknown as SessionEvent;
        expect(event.type).toBe("system.notification");
        if (event.type !== "system.notification") throw new Error("expected system.notification");
        expect(event.data.kind.type).toBe("workflow_completed");
        if (event.data.kind.type !== "workflow_completed") {
            throw new Error("expected workflow_completed");
        }
        expect(event.data.kind.workflowName).toBe("fix-ci");
        expect(event.data.kind.runId).toBe("run-1");
        expect(event.data.kind.status).toBe("completed");
        expect(event.data.kind.consumedSubagents).toBe(1);
    });
});

File: rust/tests/worker_causality_test.rs
// Copyright (c) Microsoft Corporation. All rights reserved.

#![allow(clippy::unwrap_used)]

use github_copilot_sdk::rpc::TasksSendMessageResult;
use github_copilot_sdk::session_events::{
    SystemNotification, SystemNotificationWorkflowCompletedStatus, TypedSessionEvent,
};
use serde_json::Value;

fn corpus() -> Value {
    serde_json::from_str(include_str!("../../test/worker-causality.json")).unwrap()
}

fn decode(value: &Value, event: bool) -> Value {
    if event {
        serde_json::to_value(serde_json::from_value::<TypedSessionEvent>(value.clone()).unwrap())
            .unwrap()
    } else {
        serde_json::to_value(
            serde_json::from_value::<TasksSendMessageResult>(value.clone()).unwrap(),
        )
        .unwrap()
    }
}

fn payload(value: &Value, event: bool) -> &Value {
    if event { &value["data"] } else { value }
}

fn payload_mut(value: &mut Value, event: bool) -> &mut Value {
    if event { &mut value["data"] } else { value }
}

#[test]
fn worker_causality_public_readers_preserve_source_and_product() {
    let corpus = corpus();
    for case in corpus["valid"].as_array().unwrap() {
        let event = case.get("event").is_some();
        let mut wire = case[if event { "event" } else { "result" }].clone();
        assert_eq!(
            payload(&decode(&wire, event), event)["workerCausality"],
            payload(&wire, event)["workerCausality"],
            "{}",
            case["name"]
        );
        payload_mut(&mut wire, event)
            .as_object_mut()
            .unwrap()
            .remove("workerCausality");
        let baseline = serde_json::to_vec(&decode(&wire, event)).unwrap();
        for invalid in corpus["invalid"].as_array().unwrap() {
            payload_mut(&mut wire, event)["workerCausality"] = invalid["value"].clone();
            assert_eq!(
                serde_json::to_vec(&decode(&wire, event)).unwrap(),
                baseline,
                "{}: {}",
                case["name"],
                invalid["name"]
            );
        }
    }
}

#[test]
fn worker_causality_exact_utf8_budget_and_unknown_fields() {
    let corpus = corpus();
    for case in corpus["boundaries"].as_array().unwrap() {
        let mut wire = corpus["valid"][0]["event"].clone();
        wire["data"]["workerCausality"] = case["value"].clone();
        let decoded = decode(&wire, true);
        assert_eq!(
            decoded["data"].get("workerCausality").is_some(),
            case["accepted"].as_bool().unwrap(),
            "{}",
            case["name"]
        );
        assert_eq!(decoded["data"]["content"], wire["data"]["content"]);
    }
}

#[test]
fn canonical_workflow_completion_decodes_typed_fields() {
    let corpus = corpus();
    let event: TypedSessionEvent =
        serde_json::from_value(corpus["workflowCompleted"]["event"].clone()).unwrap();
    let github_copilot_sdk::session_events::SessionEventData::SystemNotification(notification) =
        event.payload
    else {
        panic!("expected system.notification");
    };
    let SystemNotification::WorkflowCompleted(completion) = notification.kind else {
        panic!("expected workflow_completed");
    };
    assert_eq!(completion.workflow_name, "fix-ci");
    assert_eq!(completion.run_id, "run-1");
    assert!(matches!(
        completion.status,
        SystemNotificationWorkflowCompletedStatus::Completed
    ));
    assert_eq!(completion.consumed_subagents, 1);
}
